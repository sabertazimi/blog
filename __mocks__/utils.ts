let navigatorClipboardMocked = false
const writeTextMock = jest.fn()

const mockNavigatorClipboard = () => {
  if (!navigatorClipboardMocked) {
    Object.defineProperty(global.navigator, 'clipboard', {
      value: {
        writeText: writeTextMock,
      },
    })

    navigatorClipboardMocked = true
  }

  return { writeTextMock }
}

export { mockNavigatorClipboard }
