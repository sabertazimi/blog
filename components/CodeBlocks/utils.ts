import parseNumericRange from 'parse-numeric-range'
import type { Language } from 'prism-react-renderer'

const normalizeCode = (code: string = '') => code.replace(/\n+$/, '')

function isLanguage(language: string): language is Language {
  return language.length > 0
}

function normalizeLanguage(languageClass?: string): Language {
  if (languageClass != null) {
    const language = languageClass.replace('language-', '')

    if (isLanguage(language))
      return language
  }

  return 'typescript'
}

function normalizeLanguageName(language: string): string {
  switch (language) {
    case 'html':
      return 'HTML'
    case 'xml':
      return 'XML'
    case 'yml':
    case 'yaml':
      return 'YAML'
    case 'css':
      return 'CSS'
    case 'json':
      return 'JSON'
    case 'md':
    case 'markdown':
      return 'Markdown'
    case 'js':
    case 'javascript':
      return 'JavaScript'
    case 'ts':
    case 'typescript':
      return 'TypeScript'
    case 'coffee':
    case 'coffeescript':
      return 'CoffeeScript'
    case 'jsx':
    case 'tsx':
      return 'React'
    case 'objc':
    case 'objectivec':
      return 'Objective-C'
    default:
      return language.charAt(0).toUpperCase() + language.slice(1)
  }
}

function normalizeLines(expression: string): Set<number> {
  return new Set(parseNumericRange(expression))
}

export {
  normalizeCode,
  normalizeLanguage,
  normalizeLanguageName,
  normalizeLines,
}
