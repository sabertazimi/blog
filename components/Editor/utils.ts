function languageToFilepath(language?: string): string {
  switch (language) {
    case 'css':
      return '/styles.css'
    case 'js':
    case 'javascript':
      return '/App.js'
    case 'ts':
    case 'typescript':
      return '/App.ts'
    case 'jsx':
      return '/App.jsx'
    case 'tsx':
      return '/App.tsx'
    case 'vue':
      return '/src/App.vue'
    case undefined:
    default:
      return '/App.tsx'
  }
}

function normalizeFilepath(filename?: string, language?: string): string {
  if (filename == null || filename === '')
    return languageToFilepath(language)

  if (!filename.startsWith('/'))
    return `/${filename}`

  return filename
}

export { normalizeFilepath }
