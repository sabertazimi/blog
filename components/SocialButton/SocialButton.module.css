.button.button[role='link'] {
  @apply rounded-full text-sm text-light;
  @apply focus-visible:outline-dashed focus-visible:outline-2 focus-visible:outline-offset-1 focus-visible:outline-secondary;
}
