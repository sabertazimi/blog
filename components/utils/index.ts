import classNames from 'classnames';

export { classNames };
