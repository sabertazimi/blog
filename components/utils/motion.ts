export { AnimatePresence, motion } from 'framer-motion';
export type { MotionProps } from 'framer-motion';
