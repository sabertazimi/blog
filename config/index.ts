export { colors, getColorByName, socialColors } from './colors'
export type { SocialColor } from './colors'
export { routes } from './routes'
export type { Route } from './routes'
export { siteConfig } from './site'
export type { Book, SiteConfig } from './site'
export { socialList, socialQuery } from './social'
export type { SocialType } from './social'
