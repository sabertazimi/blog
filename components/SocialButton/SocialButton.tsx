import Button from '@components/Button'
import { SocialIcon } from '@components/Icons'
import { Bounce } from '@components/Motion'
import { cx } from '@components/utils'
import type { SocialType } from '@types'
import type { HTMLProps } from 'react'
import styles from './SocialButton.module.css'

interface Props extends HTMLProps<HTMLButtonElement> {
  type: SocialType
  url: string
  className?: string
  color?: string
}

function SocialButton({
  type,
  url,
  className,
  color = '',
}: Props): JSX.Element {
  return (
    <Bounce>
      <Button
        role="link"
        size="large"
        type="link"
        className={cx(styles.button, className)}
        style={color ? { backgroundColor: color } : {}}
        icon={SocialIcon(type)}
        href={`${url}`}
      />
    </Bounce>
  )
}

export default SocialButton
