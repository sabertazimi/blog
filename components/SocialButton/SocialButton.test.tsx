import { axe } from 'jest-axe'
import SocialButton from './SocialButton'
import { siteConfig } from '@/config'
import type { SocialType } from '@/types'
import { render, screen } from '@/utils'

describe('SocialButton', () => {
  const mockSocials = [...Object.keys(siteConfig.socials), 'default']

  it.each(mockSocials)(
    'should render [%s] button correctly (snapshot)',
    (social) => {
      const { container } = render(
        <SocialButton
          type={social as SocialType}
          url={`https://${social}.com`}
        />,
      )

      expect(container).toMatchSnapshot()
    },
  )

  it('should render colorful button correctly (snapshot)', () => {
    const { container } = render(
      <SocialButton type="github" url="https://github.com" color="#299954" />,
    )

    expect(container).toMatchSnapshot()
  })

  it.each(mockSocials)(
    'should render [%s] button accessibility guidelines (AXE)',
    async (social) => {
      const { container } = render(
        <SocialButton
          type={social as SocialType}
          url={`https://${social}.com`}
        />,
      )

      const a11y = await axe(container)

      expect(a11y).toHaveNoViolations()
    },
  )

  it('should render colorful button accessibility guidelines (AXE)', async () => {
    const { container } = render(
      <SocialButton type="github" url="https://github.com" color="#299954" />,
    )

    const a11y = await axe(container)

    expect(a11y).toHaveNoViolations()
  })

  it.each(mockSocials)(
    'should render [%s] button with correct structure',
    (social) => {
      render(
        <SocialButton
          type={social as SocialType}
          url={`https://${social}.com`}
        />,
      )

      const link = screen.getByRole('link')
      const icon = screen.getByRole('img')

      expect(link).toBeInTheDocument()
      expect(icon).toBeInTheDocument()
      expect(link).toContainElement(icon)
    },
  )

  it.each(mockSocials)(
    'should render [%s] button with correct URL',
    (social) => {
      render(
        <SocialButton
          type={social as SocialType}
          url={`https://${social}.com`}
        />,
      )

      const link = screen.getByRole('link')

      expect(link).toHaveAttribute('href', `https://${social}.com`)
    },
  )
})
