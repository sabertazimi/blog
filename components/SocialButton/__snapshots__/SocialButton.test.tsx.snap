// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialButton should render [default] button correctly (snapshot) 1`] = `
<div>
  <div
    class="inline-flex"
    style="opacity: 0; transform: translateX(20px) translateZ(0);"
    tabindex="0"
  >
    <a
      class="ant-btn ant-btn-link ant-btn-lg ant-btn-icon-only button"
      href="https://default.com"
      role="link"
    >
      <span
        aria-label="rocket"
        class="anticon anticon-rocket icon"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="rocket"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M864 736c0-111.6-65.4-208-160-252.9V317.3c0-15.1-5.3-29.7-15.1-41.2L536.5 95.4C530.1 87.8 521 84 512 84s-18.1 3.8-24.5 11.4L335.1 276.1a63.97 63.97 0 00-15.1 41.2v165.8C225.4 528 160 624.4 160 736h156.5c-2.3 7.2-3.5 15-3.5 23.8 0 22.1 7.6 43.7 21.4 60.8a97.2 97.2 0 0043.1 30.6c23.1 54 75.6 88.8 134.5 88.8 29.1 0 57.3-8.6 81.4-24.8 23.6-15.8 41.9-37.9 53-64a97 97 0 0043.1-30.5 97.52 97.52 0 0021.4-60.8c0-8.4-1.1-16.4-3.1-23.8H864zM762.3 621.4c9.4 14.6 17 30.3 22.5 46.6H700V558.7a211.6 211.6 0 0162.3 62.7zM388 483.1V318.8l124-147 124 147V668H388V483.1zM239.2 668c5.5-16.3 13.1-32 22.5-46.6 16.3-25.2 37.5-46.5 62.3-62.7V668h-84.8zm388.9 116.2c-5.2 3-11.2 4.2-17.1 3.4l-19.5-2.4-2.8 19.4c-5.4 37.9-38.4 66.5-76.7 66.5-38.3 0-71.3-28.6-76.7-66.5l-2.8-19.5-19.5 2.5a27.7 27.7 0 01-17.1-3.5c-8.7-5-14.1-14.3-14.1-24.4 0-10.6 5.9-19.4 14.6-23.8h231.3c8.8 4.5 14.6 13.3 14.6 23.8-.1 10.2-5.5 19.6-14.2 24.5zM464 400a48 48 0 1096 0 48 48 0 10-96 0z"
          />
        </svg>
      </span>
    </a>
  </div>
</div>
`;

exports[`SocialButton should render [facebook] button correctly (snapshot) 1`] = `
<div>
  <div
    class="inline-flex"
    style="opacity: 0; transform: translateX(20px) translateZ(0);"
    tabindex="0"
  >
    <a
      class="ant-btn ant-btn-link ant-btn-lg ant-btn-icon-only button"
      href="https://facebook.com"
      role="link"
    >
      <span
        aria-label="facebook"
        class="anticon anticon-facebook icon"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="facebook"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-32 736H663.9V602.2h104l15.6-120.7H663.9v-77.1c0-35 9.7-58.8 59.8-58.8h63.9v-108c-11.1-1.5-49-4.8-93.2-4.8-92.2 0-155.3 56.3-155.3 159.6v89H434.9v120.7h104.3V848H176V176h672v672z"
          />
        </svg>
      </span>
    </a>
  </div>
</div>
`;

exports[`SocialButton should render [github] button correctly (snapshot) 1`] = `
<div>
  <div
    class="inline-flex"
    style="opacity: 0; transform: translateX(20px) translateZ(0);"
    tabindex="0"
  >
    <a
      class="ant-btn ant-btn-link ant-btn-lg ant-btn-icon-only button"
      href="https://github.com"
      role="link"
    >
      <span
        aria-label="github"
        class="anticon anticon-github icon"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="github"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M511.6 76.3C264.3 76.2 64 276.4 64 523.5 64 718.9 189.3 885 363.8 946c23.5 5.9 19.9-10.8 19.9-22.2v-77.5c-135.7 15.9-141.2-73.9-150.3-88.9C215 726 171.5 718 184.5 703c30.9-15.9 62.4 4 98.9 57.9 26.4 39.1 77.9 32.5 104 26 5.7-23.5 17.9-44.5 34.7-60.8-140.6-25.2-199.2-111-199.2-213 0-49.5 16.3-95 48.3-131.7-20.4-60.5 1.9-112.3 4.9-120 58.1-5.2 118.5 41.6 123.2 45.3 33-8.9 70.7-13.6 112.9-13.6 42.4 0 80.2 4.9 113.5 13.9 11.3-8.6 67.3-48.8 121.3-43.9 2.9 7.7 24.7 58.3 5.5 118 32.4 36.8 48.9 82.7 48.9 132.3 0 102.2-59 188.1-200 212.9a127.5 127.5 0 0138.1 91v112.5c.8 9 0 17.9 15 17.9 177.1-59.7 304.6-227 304.6-424.1 0-247.2-200.4-447.3-447.5-447.3z"
          />
        </svg>
      </span>
    </a>
  </div>
</div>
`;

exports[`SocialButton should render [linkedin] button correctly (snapshot) 1`] = `
<div>
  <div
    class="inline-flex"
    style="opacity: 0; transform: translateX(20px) translateZ(0);"
    tabindex="0"
  >
    <a
      class="ant-btn ant-btn-link ant-btn-lg ant-btn-icon-only button"
      href="https://linkedin.com"
      role="link"
    >
      <span
        aria-label="linkedin"
        class="anticon anticon-linkedin icon"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="linkedin"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M847.7 112H176.3c-35.5 0-64.3 28.8-64.3 64.3v671.4c0 35.5 28.8 64.3 64.3 64.3h671.4c35.5 0 64.3-28.8 64.3-64.3V176.3c0-35.5-28.8-64.3-64.3-64.3zm0 736c-447.8-.1-671.7-.2-671.7-.3.1-447.8.2-671.7.3-671.7 447.8.1 671.7.2 671.7.3-.1 447.8-.2 671.7-.3 671.7zM230.6 411.9h118.7v381.8H230.6zm59.4-52.2c37.9 0 68.8-30.8 68.8-68.8a68.8 68.8 0 10-137.6 0c-.1 38 30.7 68.8 68.8 68.8zm252.3 245.1c0-49.8 9.5-98 71.2-98 60.8 0 61.7 56.9 61.7 101.2v185.7h118.6V584.3c0-102.8-22.2-181.9-142.3-181.9-57.7 0-96.4 31.7-112.3 61.7h-1.6v-52.2H423.7v381.8h118.6V604.8z"
          />
        </svg>
      </span>
    </a>
  </div>
</div>
`;

exports[`SocialButton should render [twitter] button correctly (snapshot) 1`] = `
<div>
  <div
    class="inline-flex"
    style="opacity: 0; transform: translateX(20px) translateZ(0);"
    tabindex="0"
  >
    <a
      class="ant-btn ant-btn-link ant-btn-lg ant-btn-icon-only button"
      href="https://twitter.com"
      role="link"
    >
      <span
        aria-label="twitter"
        class="anticon anticon-twitter icon"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="twitter"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M928 254.3c-30.6 13.2-63.9 22.7-98.2 26.4a170.1 170.1 0 0075-94 336.64 336.64 0 01-108.2 41.2A170.1 170.1 0 00672 174c-94.5 0-170.5 76.6-170.5 170.6 0 13.2 1.6 26.4 4.2 39.1-141.5-7.4-267.7-75-351.6-178.5a169.32 169.32 0 00-23.2 86.1c0 59.2 30.1 111.4 76 142.1a172 172 0 01-77.1-21.7v2.1c0 82.9 58.6 151.6 136.7 167.4a180.6 180.6 0 01-44.9 5.8c-11.1 0-21.6-1.1-32.2-2.6C211 652 273.9 701.1 348.8 702.7c-58.6 45.9-132 72.9-211.7 72.9-14.3 0-27.5-.5-41.2-2.1C171.5 822 261.2 850 357.8 850 671.4 850 843 590.2 843 364.7c0-7.4 0-14.8-.5-22.2 33.2-24.3 62.3-54.4 85.5-88.2z"
          />
        </svg>
      </span>
    </a>
  </div>
</div>
`;

exports[`SocialButton should render [weibo] button correctly (snapshot) 1`] = `
<div>
  <div
    class="inline-flex"
    style="opacity: 0; transform: translateX(20px) translateZ(0);"
    tabindex="0"
  >
    <a
      class="ant-btn ant-btn-link ant-btn-lg ant-btn-icon-only button"
      href="https://weibo.com"
      role="link"
    >
      <span
        aria-label="weibo"
        class="anticon anticon-weibo icon"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="weibo"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M457.3 543c-68.1-17.7-145 16.2-174.6 76.2-30.1 61.2-1 129.1 67.8 151.3 71.2 23 155.2-12.2 184.4-78.3 28.7-64.6-7.2-131-77.6-149.2zm-52 156.2c-13.8 22.1-43.5 31.7-65.8 21.6-22-10-28.5-35.7-14.6-57.2 13.7-21.4 42.3-31 64.4-21.7 22.4 9.5 29.6 35 16 57.3zm45.5-58.5c-5 8.6-16.1 12.7-24.7 9.1-8.5-3.5-11.2-13.1-6.4-21.5 5-8.4 15.6-12.4 24.1-9.1 8.7 3.2 11.8 12.9 7 21.5zm334.5-197.2c15 4.8 31-3.4 35.9-18.3 11.8-36.6 4.4-78.4-23.2-109a111.39 111.39 0 00-106-34.3 28.45 28.45 0 00-21.9 33.8 28.39 28.39 0 0033.8 21.8c18.4-3.9 38.3 1.8 51.9 16.7a54.2 54.2 0 0111.3 53.3 28.45 28.45 0 0018.2 36zm99.8-206c-56.7-62.9-140.4-86.9-217.7-70.5a32.98 32.98 0 00-25.4 39.3 33.12 33.12 0 0039.3 25.5c55-11.7 114.4 5.4 154.8 50.1 40.3 44.7 51.2 105.7 34 159.1-5.6 17.4 3.9 36 21.3 41.7 17.4 5.6 36-3.9 41.6-21.2v-.1c24.1-75.4 8.9-161.1-47.9-223.9zM729 499c-12.2-3.6-20.5-6.1-14.1-22.1 13.8-34.7 15.2-64.7.3-86-28-40.1-104.8-37.9-192.8-1.1 0 0-27.6 12.1-20.6-9.8 13.5-43.5 11.5-79.9-9.6-101-47.7-47.8-174.6 1.8-283.5 110.6C127.3 471.1 80 557.5 80 632.2 80 775.1 263.2 862 442.5 862c235 0 391.3-136.5 391.3-245 0-65.5-55.2-102.6-104.8-118zM443 810.8c-143 14.1-266.5-50.5-275.8-144.5-9.3-93.9 99.2-181.5 242.2-195.6 143-14.2 266.5 50.5 275.8 144.4C694.4 709 586 796.6 443 810.8z"
          />
        </svg>
      </span>
    </a>
  </div>
</div>
`;

exports[`SocialButton should render colorful button correctly (snapshot) 1`] = `
<div>
  <div
    class="inline-flex"
    style="opacity: 0; transform: translateX(20px) translateZ(0);"
    tabindex="0"
  >
    <a
      class="ant-btn ant-btn-link ant-btn-lg ant-btn-icon-only button"
      href="https://github.com"
      role="link"
      style="background-color: rgb(41, 153, 84);"
    >
      <span
        aria-label="github"
        class="anticon anticon-github icon"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="github"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M511.6 76.3C264.3 76.2 64 276.4 64 523.5 64 718.9 189.3 885 363.8 946c23.5 5.9 19.9-10.8 19.9-22.2v-77.5c-135.7 15.9-141.2-73.9-150.3-88.9C215 726 171.5 718 184.5 703c30.9-15.9 62.4 4 98.9 57.9 26.4 39.1 77.9 32.5 104 26 5.7-23.5 17.9-44.5 34.7-60.8-140.6-25.2-199.2-111-199.2-213 0-49.5 16.3-95 48.3-131.7-20.4-60.5 1.9-112.3 4.9-120 58.1-5.2 118.5 41.6 123.2 45.3 33-8.9 70.7-13.6 112.9-13.6 42.4 0 80.2 4.9 113.5 13.9 11.3-8.6 67.3-48.8 121.3-43.9 2.9 7.7 24.7 58.3 5.5 118 32.4 36.8 48.9 82.7 48.9 132.3 0 102.2-59 188.1-200 212.9a127.5 127.5 0 0138.1 91v112.5c.8 9 0 17.9 15 17.9 177.1-59.7 304.6-227 304.6-424.1 0-247.2-200.4-447.3-447.5-447.3z"
          />
        </svg>
      </span>
    </a>
  </div>
</div>
`;
